functionary/builder/utils.py

```diff
@@ -112,45 +112,56 @@ def build_package(build_id: UUID):
     os.makedirs(workdir)
 
     build = Build.objects.select_related("environment").get(id=build_id)
+    build.status = Build.IN_PROGRESS
+    build.save()
+
     environment = build.environment
     package_contents = build.resources.package_contents
     package_definition = build.resources.package_definition
 
-    # TODO: This sort of direct access of package_definition should be discouraged, as
-    #       it results in package schema parsing being spread throughout the code. It
-    #       should live in one place and we should call helpers in places like this.
-    name = package_definition["name"]
-    language = package_definition["language"]
-    dockerfile = f"builder/docker/{language}.Dockerfile"
-    image_name = f"{environment.id}/{name}:{build_id}"
+    image_name, dockerfile = build.resources.image_details
     full_image_name = f"{settings.REGISTRY}/{image_name}"
 
-    _extract_package_contents(package_contents, workdir)
-    _load_dockerfile_template(dockerfile, workdir)
-
-    image, build_log = docker_client.images.build(
-        path=workdir,
-        pull=True,
-        forcerm=True,
-        tag=full_image_name,
-    )
-
-    docker_client.images.push(full_image_name)
-
-    logger.debug(f"Cleaning up remnants of build {build_id}")
-    docker_client.images.remove(image.id)
-    shutil.rmtree(workdir)
-
     with transaction.atomic():
         package = _create_package_from_definition(
             package_definition, environment, image_name
         )
-        build.status = Build.COMPLETE
         build.package = package
         build.save()
```

Comment:
One of the goals I think we had for this which was maybe lost was the idea of setting the `build.package` at the time that the `Build` object gets created, rather than waiting until the build job gets picked up by a worker.  The idea being that if the build job takes some time to actually get picked up for whatever reason, right now you don't actually have a way to identify which build is yours when checking the status unless you already have the build id.

I believe this just requires doing some basic handling of the `package_definition` inside of the `initiate_build` function, where you find the corresponding package if it already exists and set that on the build.

If no matching package exists, then my initial thought was that we leave the `package` field blank rather than creating a new one, at which point the field will get populated by the actual build worker as before.  That way there isn't some unusable package left hanging around if the build fails.  That's probably the easiest thing for now, though I could see in the future creating the package and just having it not be actually usable or visible in the packages list until the build completes.

Reply:
Updated the `initiate_build` to look for the package from the `package_definition`'s name value. If it doesn't exist it'll create the package during build time and and associate it here.